### PR TITLE
feat: vehicle off-route updates

### LINKED ISSUE #118: On-the-fly vehicle off-route updates
Sometimes a vehicle can be taken off a route (e.g., if there's a defect), the map has to be updated accordingly - the vehicle should be hidden. The API has just been extended to provide such notifications in real-time:  https://github.com/roataway/api-documentation#eventroute

Summary of how it works:
- Subscribe to `event.route.<upstream route id>` (e.g. `event.route.52` for route with upstream_id=52, defined here https://github.com/roataway/infrastructure-data/blob/develop/routes.csv)
- Payloads will be a JSON: `{"event": "remove", "rtu_id": "0000115", "board": "2186"}`

The example above shows that vehicle with tracker ID `0000115`, assigned to board `2186` was removed from the route that has the id `52`.

Currently there's an undocumented attribute in the payload:  `"new_route_id": 1` - it shows the ID of the new route the vehicle has been re-assigned to. However, note that sometimes a vehicle can be taken offline for good - so this parameter will be absent. Whether it needs to be there or now - the subject is open for discussion.

Hiding the vehicle: ideally, we hide it with a "fade out" effect, to signalize to users that it didn't just vanish like a fart in the wind. The best UX for this is also an open matter.

### PR DESCRIPTION
**Pull request by strdr4605**
Close #118

Any ideas on how to receive `"remove"` event for proper testing?!!

### CODE REVIEW

**Comment by roata** [comment]
I can send a real one straight from the backend. We can do this for a pseudo-route that doesn't exist, to avoid interference with reality.

Otherwise, we'd have to somehow mock the whole thing and I am not sure how to best do that in Typescript or JavaScript.

**Comment by strdr4605** [comment]
> I can send a real one straight from the backend. We can do this for a pseudo-route that doesn't exist, to avoid interference with reality.

@roata could you do this as a review/testing process for this PR?

**Comment by roata** [comment]
@strdr4605 yes; we'd have to synchronize our schedules so we're both available at the time when you conduct your experiment. When would be a good time for you? (specify the timezone too, I'm in GMT+1)

**Comment by strdr4605** [comment]
@roata I am available tomorrow from 13:00 till 21:00 GMT+1

**Comment by iamandrewluca** [comment]
I'll postpone this PR from reviewing and merging till anyone will tag me and PR will be ready for review.

**Comment by roata** [comment]
It seems that we have forgotten about this PR, perhaps we can give it another shot. I tried it out and my conclusion is that it works, but there's a quirk in one case.

Given:
- we're focused on vehicle X, currently linked to route 30.
- the map currently displays route 30 and 32

Then:
- If X is switched from 30 to 32 -> I see a correct transition, the label and the colour of the marker change as expected, within a few moments.
- However, if it is switched to route 3 (or any other route that I am currently not viewing), then the marker remains on the map with its old colour and label. It takes about 1 minute for it to reflect the new settings.
- Another scenario I tried: when a vehicle from route 3 (or any other route I am currently not viewing) is moved to 30, it doesn't show up, unless I refresh the page (I waited for circa 4 minutes).

Considering that events are pushed immediately from the backend (I saw the message with another subscriber), it is strange that we experience such delays.

**Comment by strdr4605** [comment]
When developing this feature I did not have the context of these use cases.

---

> Another scenario I tried: when a vehicle from route 3 (or any other route I am currently not viewing) is moved to 30, it doesn't show up, unless I refresh the page (I waited for circa 4 minutes).

An idea about this issue: 
maybe **route 3** (not viewing) -> **30** trigger a _remove event_ and it's isRemoved remains true forever.

To not create any discomfort it would be nice to have temporary access to trigger these use cases while developing.

**Comment by roata** [comment]
I am not sure I understand you, but let me provide more details about how this works on the server side:

Example: vehicle X is moved from route 3 to route 30.

- the event is triggered only once, and it is published for route 3
- there is no "new vehicle added" event for route 30
- you start receiving new telemetry from this vehicle on route 30 (just like in the beginning, when you open the page, choose a route and telemetry starts flowing in and the vehicle markers show up one by one)

In other words, this problematic scenario **does not constitute a special case**; the scope of this issue is to just get it off the map for a particular route.


The temporary access request is very challenging, because the whole thing is locked down and not exposed externally in any way. There's just a public read-only stream of post-processed telemetry, everything else is inside a VPN and "behind seven proxies" :-). The easiest option is to trigger these events on-demand. The second approach is to write a tool and wrap it into a REST API that is exposed via some URL, but it would constitute more work. Perhaps we can sync over some IM tool?

**Comment by strdr4605** [comment]
After testing the first implementation, we found out that after the `"remove"` event, old boards were not deleted from position object.

**Comment by ralienpp** [comment]
@iamandrewluca, can you have a look at the logic?

My own observations are quite simple, there are several typos: `oppasity` and `opasity` should be `opacity`.

File: src/components/routes-positions.component.tsx [hunk 1]
```diff
@@ -9,12 +9,19 @@ import { useRtecClient } from '../shared/rtec-client/rtec-client.hook'
...
+import { eventRoute, EventRouteFrameBody } from '../shared/rtec-client/subscriptions/event.route'
 
 const navigationSvgPath = `M 12.037109,3.2597656 C 7.46559,3.2596004 3.7596004,6.96559 3.7597656,11.537109 c -1.655e-4,4.571519 3.7058242,8.277509 8.2773434,8.277344 4.571519,1.66e-4 8.27751,-3.705825 8.277344,-8.277344 1.65e-4,-4.5715192 -3.705825,-8.2775089 -8.277344,-8.2773434 z`
 
 const MARKER_TTL_MILLIS = 5 * 60 * 1000
 const TTL_CHECK_INTERVAL_MILLIS = 60 * 1000
 
+// Delay before board will be deleted from positions object after remove event
+const REMOVE_EVENT_TTL_MILLIS = 4000
+// Step to decrease marker oppasity from 1 to 0
```
Comment (strdr4605) [inline src/components/routes-positions.component.tsx:21]:
```suggestion
// Step to decrease marker opacity from 1 to 0
```

File: src/components/routes-positions.component.tsx [hunk 2]
```diff
@@ -157,6 +213,22 @@ function TransportMarker(props: TransportMarkerProps) {
   const { t } = useTranslation()
   const { colors } = useRouteColors()
 
+  const opasityRef = useRef<number>(1)
```
Comment (strdr4605) [inline src/components/routes-positions.component.tsx:216]:
```suggestion
  const opacityRef = useRef<number>(1)
```

File: src/components/routes-positions.component.tsx [hunk 3]
```diff
@@ -157,6 +213,22 @@ function TransportMarker(props: TransportMarkerProps) {
...
+  useEffect(() => {
+    if (!transport.isRemoved) {
+      return
+    }
+
+    const timerId = setInterval(() => {
+      if (opasityRef.current <= 0) {
```
Comment (strdr4605) [inline src/components/routes-positions.component.tsx:223]:
```suggestion
      if (opacityRef.current <= 0) {
```

File: src/components/routes-positions.component.tsx [hunk 4]
```diff
@@ -157,6 +213,22 @@ function TransportMarker(props: TransportMarkerProps) {
...
+        clearInterval(timerId)
+      }
+      opasityRef.current -= FADE_OUT_STEP
```
Comment (strdr4605) [inline src/components/routes-positions.component.tsx:226]:
```suggestion
      opacityRef.current -= FADE_OUT_STEP
```

File: src/components/routes-positions.component.tsx [hunk 5]
```diff
@@ -188,11 +260,12 @@ function TransportMarker(props: TransportMarkerProps) {
   const vehicle = board ? boardToVehicle.get(board) : undefined
   const accessibility = vehicle && vehicle.accessibility ? ' ♿' : undefined
 
-  return (
+  return opasityRef.current > 0 ? (
```
Comment (strdr4605) [inline src/components/routes-positions.component.tsx:263]:
```suggestion
  return opacityRef.current > 0 ? (
```

File: src/components/routes-positions.component.tsx [hunk 6]
```diff
@@ -188,11 +260,12 @@ function TransportMarker(props: TransportMarkerProps) {
...
     <Marker
       key={transport.board}
       title={transport.board}
       position={[transport.latitude, transport.longitude]}
+      opacity={opasityRef.current}
```
Comment (strdr4605) [inline src/components/routes-positions.component.tsx:268]:
```suggestion
      opacity={opacityRef.current}
```

**Comment by iamandrewluca** [comment]
I'll take a look this week.

**Review by iamandrewluca — APPROVED** [review]
🚀

**Pull request merged**